elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py

```diff
@@ -0,0 +1,702 @@
+            self.client.addSection(section)
+        if alert.path:
+            section = cardsection()
+            section.activityTitle("*Path*")
+            section.activityText(f"`{alert.path}`")
+            self.client.addSection(section)
+
+    def _get_snapshot_template(self, alert: ModelAlertModel, *args, **kwargs):
+        title = f"{self._get_display_name(alert.status)}: {alert.summary}"
+        subtitle = self._get_alert_sub_title(alert)
+
+        model_runs_report_link = get_model_runs_link(
+            alert.report_url, alert.model_unique_id
+        )
+        if model_runs_report_link:
+            action = self._get_potential_action(model_runs_report_link)
+            self.client.addPotentialAction(action)
+
+        self.client.title(title)
+        self.client.text(subtitle)
+
+        if TAGS_FIELD in (alert.alert_fields or DEFAULT_ALERT_FIELDS):
+            tags = prettify_and_dedup_list(alert.tags or [])
+            section = cardsection()
+            section.activityTitle("*Tags*")
+            section.activityText(f'_{tags or "No tags"}_')
+            self.client.addSection(section)
+
+        if OWNERS_FIELD in (alert.alert_fields or DEFAULT_ALERT_FIELDS):
+            owners = prettify_and_dedup_list(alert.owners or [])
+            section = cardsection()
+            section.activityTitle("*Owners*")
+            section.activityText(f'_{owners or "No owners"}_')
+            self.client.addSection(section)
+
+        if SUBSCRIBERS_FIELD in (alert.alert_fields or DEFAULT_ALERT_FIELDS):
+            subscribers = prettify_and_dedup_list(alert.subscribers or [])
+            section = cardsection()
+            section.activityTitle("*Subscribers*")
+            section.activityText(f'_{subscribers or "No subscribers"}_')
+            self.client.addSection(section)
+
+        if alert.message:
+            section = cardsection()
+            section.activityTitle("*Result message*")
+            section.activityText(f"```{alert.message.strip()}```")
+            self.client.addSection(section)
+
+        if alert.original_path:
+            section = cardsection()
+            section.activityTitle("*Path*")
+            section.activityText(f"`{alert.original_path}`")
+            self.client.addSection(section)
+
+    def _get_source_freshness_template(
+        self, alert: SourceFreshnessAlertModel, *args, **kwargs
+    ):
+        title = f"{self._get_display_name(alert.status)}: {alert.summary}"
+        subtitle = self._get_alert_sub_title(alert)
+
+        test_runs_report_link = get_test_runs_link(
+            alert.report_url, alert.source_freshness_execution_id
+        )
+        if test_runs_report_link:
+            action = self._get_potential_action(test_runs_report_link)
+            self.client.addPotentialAction(action)
+
+        self.client.title(title)
+        self.client.text(subtitle)
+
+        if TAGS_FIELD in (alert.alert_fields or DEFAULT_ALERT_FIELDS):
```

Comment:
Just a style comment, and I know we have that issue in the slack integration as well, but we can have some more code reuse here IMO, some generic methods that check whether a section (tags/owners/subscribers/etc) needs to be added, and adds it if it does.

Reply:
Yeah I thought about it, but in that way it would turn into something that is harder to double check with the Slack code and if I didn't make any fuckups. 

I also thought that it would make sense to refactor the slack and teams shared functionality into a base class and reuse it here. But that would require a lot more work and therefor chose not to do so. Is it OK to leave this for now?

Reply:
yeah I think so.
I want to try to separate the formatting as much as possible from any specific integration but it's hard because each has different formatting options and limitations. We'll do that sometime.

Reply:
As mentioned in slack as well, I have some refactoring work done, but now mypy returns some issues:

```
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:226: error: Item "ModelAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "error_message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:226: error: Item "TestAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:226: error: Item "SourceFreshnessAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:230: error: Item "ModelAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "error_message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:231: error: Item "ModelAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "error_message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:232: error: Item "TestAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:232: error: Item "SourceFreshnessAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:233: error: Item "TestAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "message"  [union-attr]
elementary/monitor/data_monitoring/alerts/integrations/teams/teams.py:233: error: Item "SourceFreshnessAlertModel" of "Union[TestAlertModel, ModelAlertModel, SourceFreshnessAlertModel]" has no attribute "message"  [union-attr]
```

This happens for example when I create a def like:

```py
    def _add_report_link_if_applicable(
        self,
        alert: Union[
            TestAlertModel,
            ModelAlertModel,
            SourceFreshnessAlertModel,
            GroupedByTableAlerts,
        ],
    ):

        if isinstance(alert, ModelAlertModel) or isinstance(
            alert, SourceFreshnessAlertModel
        ):
            reportlink = get_model_runs_link(alert.report_url, alert.model_unique_id)
        elif isinstance(alert, TestAlertModel):
```

Not sure if it makes sense to add `error_message` to the `AlertModel` class. WDYT?

Reply:
You can ignore my above comment! I have refactored what I think is reasonable. 

Note that I did not chose to make `_get_group_by_table_template` dependent on the `alert.alert_fields` since this is currently also not the case in Slack.

Please let me know if this is as expected now?